actors/market/tests/market_actor_test.rs

```diff
@@ -1278,6 +1279,295 @@ fn active_deals_multiple_times_with_different_providers() {
+        )
+        .unwrap_err()
+        .exit_code()
+    );
+    rt.verify();
+    // TODO: actor.checkState(rt)
+}
+
+#[cfg(test)]
+mod publish_storage_deals_failures {
+    use super::*;
+
+    #[test]
+    fn deal_end_after_deal_start() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.start_epoch = 10;
+            d.end_epoch = 9;
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn current_epoch_greater_than_start_epoch() {
+        let current_epoch = ChainEpoch::from(5);
+
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.start_epoch = current_epoch - 1;
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn deal_duration_greater_than_max_deal_duration() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.start_epoch = ChainEpoch::from(10);
+            d.end_epoch = d.start_epoch + (540 * EPOCHS_IN_DAY) + 1
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn negative_price_per_epoch() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.storage_price_per_epoch = TokenAmount::from(-1);
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn price_per_epoch_greater_than_total_filecoin() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.storage_price_per_epoch = TOTAL_FILECOIN.clone() + 1;
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn negative_provider_collateral() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.provider_collateral = TokenAmount::from(-1);
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn provider_collateral_greater_than_max_collateral() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.provider_collateral = TOTAL_FILECOIN.clone() + 1;
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn provider_collateral_less_than_bound() {
+        let f = |rt: &mut MockRuntime, d: &mut DealProposal| {
+            // with these two equal provider collatreal min is 5/100 * deal size
+            let power = StoragePower::from_i128(1 << 50).unwrap();
+            rt.set_circulating_supply(power);
+            let deal_size = 2048; // default deal size used
+            let policy = Policy::default();
+            let provider_min = (deal_size * policy.prov_collateral_percent_supply_num as u64)
+                / (policy.prov_collateral_percent_supply_denom as u64);
+            d.provider_collateral = TokenAmount::from(provider_min - 1);
```

Comment:
To improve upon the go test I think it's just as good and a lot easier read to instead use `deal_provider_collateral_bounds`.  Less important for clarity but you could also use this function for the above case.

Reply:
You'll need to expose that function as public but I think that's not an issue.